src/webgpu/api/validation/encoding/programmable/pipeline_bind_group_compat.spec.ts

```diff
@@ -502,10 +272,136 @@ g.test('it_is_invalid_to_draw_in_a_render_pass_with_missing_bind_groups')
     if (setBindGroup2) {
       renderPass.setBindGroup(1, bindGroup1);
     }
-    renderPass.setVertexBuffer(0, verticesBuffer);
+    renderPass.setVertexBuffer(0, vertexBuffer);
     renderPass.draw(3);
     renderPass.endPass();
     t.expectValidationError(() => {
       commandEncoder.finish();
     }, !_success);
   });
+
+g.test('buffer_binding_layout_should_be_exactly_equivalent,render_pipeline')
+  .desc(
+    `
+  The GPUBufferBindingLayout bindings configure should be exactly
+  same in PipelineLayout and bindgroup.
+  - TODO: test more draw functions, e.g. indirect
+  - TODO: test more visibilities, e.g. vetex
+  - TODO: bind group should be created with different layout
+  `
+  )
+  .params(u => u.combine('type', kBufferBindingTypes))
+  .fn(async t => {
+    const { type } = t.params;
+
+    // Create fixed bindGroup
+    const uniformBuffer = t.getUniformBuffer();
+
+    const bindGroup = t.device.createBindGroup({
+      entries: [
+        {
+          binding: 0,
+          resource: {
+            buffer: uniformBuffer,
+          },
+        },
+      ],
+      layout: t.device.createBindGroupLayout({
+        entries: [
+          {
+            binding: 0,
+            visibility: GPUShaderStage.FRAGMENT,
+            buffer: {}, // default type: uniform
+          },
+        ],
+      }),
+    });
+
+    // Create pipeline with different layouts
+    const pipeline = t.createRenderPipelineWithLayout(t.device, [
+      [
+        {
+          binding: 0,
+          visibility: GPUShaderStage.FRAGMENT,
+          buffer: {
+            type,
+          },
+        },
+      ],
+    ]);
+
+    const success = type === undefined || type === 'uniform';
+
+    const vertexBuffer = t.getVertexBuffer();
+
+    const commandEncoder = t.device.createCommandEncoder();
+    const renderPass = t.beginRenderPass(commandEncoder);
+    renderPass.setPipeline(pipeline);
+    renderPass.setBindGroup(0, bindGroup);
+    renderPass.setVertexBuffer(0, vertexBuffer);
+    renderPass.draw(3);
+    renderPass.endPass();
+    t.expectValidationError(() => {
+      commandEncoder.finish();
+    }, !success);
+  });
+
+g.test('sampler_binding_layout_should_be_exactly_equivalent,render_pipeline')
+  .desc(
+    `
+  The GPUSamplerBindingLayout bindings configure should be exactly
+  same in PipelineLayout and bindgroup.
+  - TODO: test more draw functions, e.g. indirect
+  - TODO: test more visibilities, e.g. vetex
+  - TODO: bind group should be created with different layout
+  `
+  )
+  .params(u => u.combine('type', kSamplerBindingTypes))
```

Comment:
this would be very easy to parameterize right now - the createSampler call doesn't even have to change.

```
  .params(u => u.combine('bglType', kSamplerBindingTypes))
  .params(u => u.combine('bgType', kSamplerBindingTypes))
```